// @dydxprotocol/starkex-eth/build/src/modules/FactRegistry.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FactRegistry = void 0;
const web3_1 = __importDefault(require("web3"));
const Constants_1 = require("../lib/Constants");
const ContractCallHelpers_1 = require("../lib/ContractCallHelpers");
class FactRegistry {
    constructor(contracts) {
        this.contracts = contracts;
    }
    getAddress() {
        return this.contracts.factRegistry.options.address;
    }
    getTransferEthFact({ recipient, humanAmount, salt, }) {
        return this.getTransferErc20Fact({
            tokenAddress: Constants_1.ZERO_ADDRESS,
            tokenDecimals: 18,
            recipient,
            humanAmount,
            salt,
        });
    }
    getTransferErc20Fact({ recipient, tokenAddress, tokenDecimals, humanAmount, salt, }) {
        const result = web3_1.default.utils.soliditySha3({ type: 'address', value: recipient }, { type: 'uint256', value: (0, ContractCallHelpers_1.humanTokenAmountToUint256)(humanAmount, tokenDecimals) }, { type: 'address', value: tokenAddress }, { type: 'uint256', value: (0, ContractCallHelpers_1.bignumberableToUint256)(salt) });
        return result;
    }
    async transferEth({ recipient, humanAmount, salt, }, options) {
        return this.contracts.send(this.contracts.factRegistry, this.contracts.factRegistry.methods.transfer(recipient, (0, ContractCallHelpers_1.bignumberableToUint256)(salt)), {
            ...options,
            value: (0, ContractCallHelpers_1.humanEthAmountToUint256)(humanAmount),
        });
    }
    async transferERC20({ recipient, tokenAddress, tokenDecimals, humanAmount, salt, }, options) {
        return this.contracts.send(this.contracts.factRegistry, this.contracts.factRegistry.methods.transferERC20(recipient, tokenAddress, (0, ContractCallHelpers_1.humanTokenAmountToUint256)(humanAmount, tokenDecimals), (0, ContractCallHelpers_1.bignumberableToUint256)(salt)), options);
    }
    async transferCollateralToken({ recipient, humanAmount, salt, }, options) {
        return this.contracts.send(this.contracts.factRegistry, this.contracts.factRegistry.methods.transferERC20(recipient, this.contracts.collateralToken.options.address, (0, ContractCallHelpers_1.humanCollateralAmountToUint256)(humanAmount), (0, ContractCallHelpers_1.bignumberableToUint256)(salt)), options);
    }
    // ============ Getters ============
    async indentify(options) {
        const result = await this.contracts.call(this.contracts.factRegistry.methods.indentify(), options);
        return result;
    }
    async hasRegisteredFact(options) {
        const result = await this.contracts.call(this.contracts.factRegistry.methods.hasRegisteredFact(), options);
        return result;
    }
    async isValid({ fact, }, options) {
        const result = await this.contracts.call(this.contracts.factRegistry.methods.isValid(fact), options);
        return result;
    }
}
exports.FactRegistry = FactRegistry;
