// @dydxprotocol/starkex-eth/build/src/lib/Constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.USDC_EXCHANGE_ADDRESSES = exports.USDC_ADDRESSES = exports.ZERO_ADDRESS = exports.COLLATERAL_ASSET_ID = exports.INTEGERS = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const usdc_abi_json_1 = __importDefault(require("../contracts/usdc-abi.json"));
const usdc_exchange_proxy_abi_json_1 = __importDefault(require("../contracts/usdc-exchange-proxy-abi.json"));
const types_1 = require("../types");
const BytesHelper_1 = require("./BytesHelper");
exports.INTEGERS = {
    ZERO: new bignumber_js_1.default(0),
    ONE: new bignumber_js_1.default(1),
    ONES_255: new bignumber_js_1.default('115792089237316195423570985008687907853269984665640564039457584007913129639935'), // 2**256-1
};
exports.COLLATERAL_ASSET_ID = {
    [types_1.Networks.MAINNET]: (0, BytesHelper_1.getAssetId)(usdc_abi_json_1.default.networks[1].address),
    [types_1.Networks.ROPSTEN]: (0, BytesHelper_1.getAssetId)(usdc_abi_json_1.default.networks[3].address),
    [types_1.Networks.GOERLI]: (0, BytesHelper_1.getAssetId)(usdc_abi_json_1.default.networks[5].address),
};
exports.ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
exports.USDC_ADDRESSES = {
    [types_1.Networks.MAINNET]: usdc_abi_json_1.default.networks[1].address,
    [types_1.Networks.ROPSTEN]: usdc_abi_json_1.default.networks[3].address,
    [types_1.Networks.GOERLI]: usdc_abi_json_1.default.networks[5].address,
};
exports.USDC_EXCHANGE_ADDRESSES = {
    [types_1.Networks.MAINNET]: usdc_exchange_proxy_abi_json_1.default.networks[1].address,
    [types_1.Networks.ROPSTEN]: usdc_exchange_proxy_abi_json_1.default.networks[3].address,
    [types_1.Networks.GOERLI]: usdc_exchange_proxy_abi_json_1.default.networks[5].address,
};
