// @dydxprotocol/starkex-eth/build/src/lib/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeResponse = exports.sendGaslessTransaction = exports.getUsdcAddress = exports.OUTCOMES = void 0;
const Constants_1 = require("./Constants");
var OUTCOMES;
(function (OUTCOMES) {
    OUTCOMES[OUTCOMES["INITIAL"] = 0] = "INITIAL";
    OUTCOMES[OUTCOMES["RESOLVED"] = 1] = "RESOLVED";
    OUTCOMES[OUTCOMES["REJECTED"] = 2] = "REJECTED";
})(OUTCOMES = exports.OUTCOMES || (exports.OUTCOMES = {}));
function getUsdcAddress(networkId) {
    const buyTokenAddress = Constants_1.USDC_ADDRESSES[networkId];
    if (buyTokenAddress === undefined) {
        throw new Error(`No buyTokenAddress with networkId: ${networkId}`);
    }
    return buyTokenAddress;
}
exports.getUsdcAddress = getUsdcAddress;
async function sendGaslessTransaction(contractCall) {
    let hashOutcome = OUTCOMES.INITIAL;
    const transactionHash = await new Promise((resolve, reject) => {
        contractCall.on('error', (error) => {
            // has not reached a transactionHash or error yet
            if (hashOutcome === OUTCOMES.INITIAL) {
                hashOutcome = OUTCOMES.REJECTED;
                reject(error);
                contractCall.off();
            }
        });
        contractCall.on('transactionHash', (txHash) => {
            // has not reached a transactionHash or error yet
            if (hashOutcome === OUTCOMES.INITIAL) {
                hashOutcome = OUTCOMES.RESOLVED;
                resolve(txHash);
                contractCall.off();
            }
        });
    });
    return normalizeResponse({ transactionHash });
}
exports.sendGaslessTransaction = sendGaslessTransaction;
function normalizeResponse(txResult) {
    const txHash = txResult.transactionHash;
    if (txHash) {
        const { transactionHash: internalHash, nonce: internalNonce, } = txHash;
        if (internalHash) {
            txResult.transactionHash = internalHash;
        }
        if (internalNonce) {
            txResult.nonce = internalNonce;
        }
    }
    return txResult;
}
exports.normalizeResponse = normalizeResponse;
