// @dydxprotocol/v3-client/build/src/modules/onboarding.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const starkex_lib_1 = require("@dydxprotocol/starkex-lib");
const eth_signing_1 = require("../eth-signing");
const helpers_1 = require("../eth-signing/helpers");
const request_helpers_1 = require("../helpers/request-helpers");
const axios_1 = require("../lib/axios");
const types_1 = require("../types");
const KEY_DERIVATION_SUPPORTED_SIGNING_METHODS = [
    types_1.SigningMethod.TypedData,
    types_1.SigningMethod.MetaMask,
    types_1.SigningMethod.MetaMaskLatest,
    types_1.SigningMethod.CoinbaseWallet,
    types_1.SigningMethod.Personal,
];
class Onboarding {
    constructor(host, web3, networkId) {
        this.host = host;
        this.networkId = networkId;
        this.signer = new eth_signing_1.SignOnboardingAction(web3, networkId);
    }
    // ============ Request Helpers ============
    async post(endpoint, data, ethereumAddress, signature = null, signingMethod = types_1.SigningMethod.TypedData) {
        const message = { action: types_1.OnboardingActionString.ONBOARDING };
        // On mainnet, include an extra onlySignOn parameter.
        if (this.networkId === 1) {
            message.onlySignOn = 'https://trade.dydx.exchange';
        }
        const url = `/v3/${endpoint}`;
        return (0, axios_1.axiosRequest)({
            url: `${this.host}${url}`,
            method: axios_1.RequestMethod.POST,
            data,
            headers: {
                'DYDX-SIGNATURE': signature || await this.signer.sign(ethereumAddress, signingMethod, message),
                'DYDX-ETHEREUM-ADDRESS': ethereumAddress,
            },
        });
    }
    /**
     * Sign the 'key derivation' onboarding message to receive a signature.
     *
     * @param ethereumAddress of the account
     * @param signingMethod Method to use for signing
     * @returns Signature used to derive your STARK key pairs
     */
    async signStarkKeyDerivationMessage(ethereumAddress, signingMethod = types_1.SigningMethod.TypedData) {
        if (!KEY_DERIVATION_SUPPORTED_SIGNING_METHODS.includes(signingMethod)) {
            throw new Error('Unsupported signing method for API key derivation');
        }
        const message = {
            action: types_1.OnboardingActionString.KEY_DERIVATION,
        };
        // On mainnet, include an extra onlySignOn parameter.
        if (this.networkId === 1) {
            message.onlySignOn = 'https://trade.dydx.exchange';
        }
        return this.signer.sign(ethereumAddress, signingMethod, message);
    }
    // ============ Requests ============
    /**
     * @description create a user, account and apiKey in one onboarding request
     *
     * @param {
     * @starkKey is the unique public key for starkwareLib operations used in the future
     * @starkKeyYCoordinate is the Y Coordinate of the unique public key for starkwareLib
     * operations used in the future
     * }
     * @param ethereumAddress of the account
     * @param signature validating the request
     * @param signingMethod for the request
     * @param referredByAffiliateLink of affiliate who referred the user
     * @param country for the user (ISO 3166-1 Alpha-2 Compliant)
     */
    async createUser(params, ethereumAddress, signature = null, signingMethod) {
        return this.post('onboarding', params, ethereumAddress, signature, signingMethod);
    }
    // ============ Key Derivation ============
    /**
     * @description Derive a STARK key pair deterministically from an Ethereum key.
     *
     * This is used by the frontend app to derive the STARK key pair in a way that is recoverable.
     * Programmatic traders may optionally derive their STARK key pair in the same way.
     *
     * @param ethereumAddress Ethereum address of the account to use for signing.
     * @param signingMethod Method to use for signing.
     */
    async deriveStarkKey(ethereumAddress, signingMethod = types_1.SigningMethod.TypedData) {
        const signature = await this.signStarkKeyDerivationMessage(ethereumAddress, signingMethod);
        return (0, starkex_lib_1.keyPairFromData)(Buffer.from((0, helpers_1.stripHexPrefix)(signature), 'hex'));
    }
    /**
     * @description Derive four STARK key pairs deterministically from an Ethereum key, with three
     * of the STARK key pairs using a signature that has had either their 'v' value, 't' value, or
     * both values rotated.
     *
     * This is used by the frontend app to derive the two STARK key pairs to ensure there will be
     * no STARK key mismatch due to a signature's 'v' or 't' value.
     *
     * @param ethereumAddress Ethereum address of the account to use for signing.
     * @param signingMethod Method to use for signing.
     */
    async deriveAllStarkKeys(ethereumAddress, signingMethod = types_1.SigningMethod.TypedData) {
        const signature = await this.signStarkKeyDerivationMessage(ethereumAddress, signingMethod);
        const rotatedSignatures = (0, helpers_1.getAllSignatureRotations)(signature);
        return rotatedSignatures.map((rotatedSignature) => (0, starkex_lib_1.keyPairFromData)(Buffer.from((0, helpers_1.stripHexPrefix)(rotatedSignature), 'hex')));
    }
    /**
     * @description Derive an API key pair deterministically from an Ethereum key.
     *
     * This is used by the frontend app to recover the default API key credentials.
     *
     * @param ethereumAddress Ethereum address of the account to use for signing.
     * @param signingMethod Method to use for signing.
     */
    async recoverDefaultApiCredentials(ethereumAddress, signingMethod = types_1.SigningMethod.TypedData) {
        if (!KEY_DERIVATION_SUPPORTED_SIGNING_METHODS.includes(signingMethod)) {
            throw new Error('Unsupported signing method for API key derivation');
        }
        const message = { action: types_1.OnboardingActionString.ONBOARDING };
        // On mainnet, include an extra onlySignOn parameter.
        if (this.networkId === 1) {
            message.onlySignOn = 'https://trade.dydx.exchange';
        }
        const signature = await this.signer.sign(ethereumAddress, signingMethod, message);
        const buffer = Buffer.from((0, helpers_1.stripHexPrefix)(signature), 'hex');
        // Get secret.
        const rBuffer = buffer.slice(0, 32);
        const rHashedData = (0, request_helpers_1.keccak256Buffer)(rBuffer);
        const secret = rHashedData.slice(0, 30);
        // Get key and passphrase.
        const sBuffer = buffer.slice(32, 64);
        const sHashedData = (0, request_helpers_1.keccak256Buffer)(sBuffer);
        const key = sHashedData.slice(0, 16);
        const passphrase = sHashedData.slice(16, 31);
        return {
            secret: toBase64Url(secret),
            key: uuidFormatKey(key),
            passphrase: toBase64Url(passphrase),
        };
    }
}
exports.default = Onboarding;
function uuidFormatKey(keyBuffer) {
    const key = keyBuffer.toString('hex');
    return [
        key.slice(0, 8),
        key.slice(8, 12),
        key.slice(12, 16),
        key.slice(16, 20),
        key.slice(20, 32),
    ].join('-');
}
function toBase64Url(base64) {
    return base64.toString('base64')
        .replace(/=/g, '')
        .replace(/\+/g, '-')
        .replace(/\//g, '_');
}
