// @dydxprotocol/starkex-lib/build/src/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ASSET_QUANTUM_SIZE = exports.SYNTHETIC_ASSET_ID_MAP = exports.COLLATERAL_ASSET_ID_BY_NETWORK_ID = exports.COLLATERAL_ASSET_ADDRESS_BY_NETWORK = exports.ASSET_RESOLUTION = exports.SYNTHETIC_ASSET_MAP = exports.SYNTHETIC_ASSETS = exports.COLLATERAL_ASSET = exports.ALL_ASSETS = void 0;
const bn_js_1 = __importDefault(require("bn.js"));
const keccak_1 = require("ethereum-cryptography/keccak");
const lodash_1 = __importDefault(require("lodash"));
const util_1 = require("./lib/util");
const types_1 = require("./types");
exports.ALL_ASSETS = Object.values(types_1.DydxAsset);
exports.COLLATERAL_ASSET = types_1.DydxAsset.USDC;
exports.SYNTHETIC_ASSETS = lodash_1.default.without(exports.ALL_ASSETS, exports.COLLATERAL_ASSET);
/**
 * Mapping from a dYdX market to the synthetic asset for that market.
 */
exports.SYNTHETIC_ASSET_MAP = {
    [types_1.DydxMarket.BTC_USD]: types_1.DydxAsset.BTC,
    [types_1.DydxMarket.ETH_USD]: types_1.DydxAsset.ETH,
    [types_1.DydxMarket.LINK_USD]: types_1.DydxAsset.LINK,
    [types_1.DydxMarket.AAVE_USD]: types_1.DydxAsset.AAVE,
    [types_1.DydxMarket.UNI_USD]: types_1.DydxAsset.UNI,
    [types_1.DydxMarket.SUSHI_USD]: types_1.DydxAsset.SUSHI,
    [types_1.DydxMarket.SOL_USD]: types_1.DydxAsset.SOL,
    [types_1.DydxMarket.YFI_USD]: types_1.DydxAsset.YFI,
    [types_1.DydxMarket.ONEINCH_USD]: types_1.DydxAsset.ONEINCH,
    [types_1.DydxMarket.AVAX_USD]: types_1.DydxAsset.AVAX,
    [types_1.DydxMarket.SNX_USD]: types_1.DydxAsset.SNX,
    [types_1.DydxMarket.CRV_USD]: types_1.DydxAsset.CRV,
    [types_1.DydxMarket.UMA_USD]: types_1.DydxAsset.UMA,
    [types_1.DydxMarket.DOT_USD]: types_1.DydxAsset.DOT,
    [types_1.DydxMarket.DOGE_USD]: types_1.DydxAsset.DOGE,
    [types_1.DydxMarket.MATIC_USD]: types_1.DydxAsset.MATIC,
    [types_1.DydxMarket.MKR_USD]: types_1.DydxAsset.MKR,
    [types_1.DydxMarket.FIL_USD]: types_1.DydxAsset.FIL,
    [types_1.DydxMarket.ADA_USD]: types_1.DydxAsset.ADA,
    [types_1.DydxMarket.ATOM_USD]: types_1.DydxAsset.ATOM,
    [types_1.DydxMarket.COMP_USD]: types_1.DydxAsset.COMP,
    [types_1.DydxMarket.BCH_USD]: types_1.DydxAsset.BCH,
    [types_1.DydxMarket.LTC_USD]: types_1.DydxAsset.LTC,
    [types_1.DydxMarket.EOS_USD]: types_1.DydxAsset.EOS,
    [types_1.DydxMarket.ALGO_USD]: types_1.DydxAsset.ALGO,
    [types_1.DydxMarket.ZRX_USD]: types_1.DydxAsset.ZRX,
    [types_1.DydxMarket.XMR_USD]: types_1.DydxAsset.XMR,
    [types_1.DydxMarket.ZEC_USD]: types_1.DydxAsset.ZEC,
    [types_1.DydxMarket.ENJ_USD]: types_1.DydxAsset.ENJ,
    [types_1.DydxMarket.ETC_USD]: types_1.DydxAsset.ETC,
    [types_1.DydxMarket.XLM_USD]: types_1.DydxAsset.XLM,
    [types_1.DydxMarket.TRX_USD]: types_1.DydxAsset.TRX,
    [types_1.DydxMarket.XTZ_USD]: types_1.DydxAsset.XTZ,
    [types_1.DydxMarket.HNT_USD]: types_1.DydxAsset.HNT,
    [types_1.DydxMarket.ICP_USD]: types_1.DydxAsset.ICP,
    [types_1.DydxMarket.RUNE_USD]: types_1.DydxAsset.RUNE,
    [types_1.DydxMarket.LUNA_USD]: types_1.DydxAsset.LUNA,
    [types_1.DydxMarket.NEAR_USD]: types_1.DydxAsset.NEAR,
    [types_1.DydxMarket.AR_USD]: types_1.DydxAsset.AR,
    [types_1.DydxMarket.FLOW_USD]: types_1.DydxAsset.FLOW,
    [types_1.DydxMarket.PERP_USD]: types_1.DydxAsset.PERP,
    [types_1.DydxMarket.REN_USD]: types_1.DydxAsset.REN,
    [types_1.DydxMarket.CELO_USD]: types_1.DydxAsset.CELO,
    [types_1.DydxMarket.KSM_USD]: types_1.DydxAsset.KSM,
    [types_1.DydxMarket.BAL_USD]: types_1.DydxAsset.BAL,
    [types_1.DydxMarket.BNT_USD]: types_1.DydxAsset.BNT,
    [types_1.DydxMarket.MIR_USD]: types_1.DydxAsset.MIR,
    [types_1.DydxMarket.SRM_USD]: types_1.DydxAsset.SRM,
    [types_1.DydxMarket.LON_USD]: types_1.DydxAsset.LON,
    [types_1.DydxMarket.DODO_USD]: types_1.DydxAsset.DODO,
    [types_1.DydxMarket.ALPHA_USD]: types_1.DydxAsset.ALPHA,
    [types_1.DydxMarket.WNXM_USD]: types_1.DydxAsset.WNXM,
    [types_1.DydxMarket.XCH_USD]: types_1.DydxAsset.XCH,
};
/**
 * The resolution represents the number of decimals of precision used in the Starkware system.
 *
 * For example, a resolution of 9 for ETH means that 1e-9 ETH = 1 Gwei is the smallest unit.
 */
exports.ASSET_RESOLUTION = {
    [types_1.DydxAsset.USDC]: 6,
    [types_1.DydxAsset.BTC]: 10,
    [types_1.DydxAsset.ETH]: 9,
    [types_1.DydxAsset.LINK]: 7,
    [types_1.DydxAsset.AAVE]: 8,
    [types_1.DydxAsset.UNI]: 7,
    [types_1.DydxAsset.SUSHI]: 7,
    [types_1.DydxAsset.SOL]: 7,
    [types_1.DydxAsset.YFI]: 10,
    [types_1.DydxAsset.ONEINCH]: 7,
    [types_1.DydxAsset.AVAX]: 7,
    [types_1.DydxAsset.SNX]: 7,
    [types_1.DydxAsset.CRV]: 6,
    [types_1.DydxAsset.UMA]: 7,
    [types_1.DydxAsset.DOT]: 7,
    [types_1.DydxAsset.DOGE]: 5,
    [types_1.DydxAsset.MATIC]: 6,
    [types_1.DydxAsset.MKR]: 9,
    [types_1.DydxAsset.FIL]: 7,
    [types_1.DydxAsset.ADA]: 6,
    [types_1.DydxAsset.ATOM]: 7,
    [types_1.DydxAsset.COMP]: 8,
    [types_1.DydxAsset.BCH]: 8,
    [types_1.DydxAsset.LTC]: 8,
    [types_1.DydxAsset.EOS]: 6,
    [types_1.DydxAsset.ALGO]: 6,
    [types_1.DydxAsset.ZRX]: 6,
    [types_1.DydxAsset.XMR]: 8,
    [types_1.DydxAsset.ZEC]: 8,
    [types_1.DydxAsset.ENJ]: 6,
    [types_1.DydxAsset.ETC]: 7,
    [types_1.DydxAsset.XLM]: 5,
    [types_1.DydxAsset.TRX]: 4,
    [types_1.DydxAsset.XTZ]: 6,
    [types_1.DydxAsset.HNT]: 7,
    [types_1.DydxAsset.ICP]: 7,
    [types_1.DydxAsset.RUNE]: 6,
    [types_1.DydxAsset.LUNA]: 6,
    [types_1.DydxAsset.NEAR]: 6,
    [types_1.DydxAsset.AR]: 7,
    [types_1.DydxAsset.FLOW]: 7,
    [types_1.DydxAsset.PERP]: 6,
    [types_1.DydxAsset.REN]: 5,
    [types_1.DydxAsset.CELO]: 6,
    [types_1.DydxAsset.KSM]: 8,
    [types_1.DydxAsset.BAL]: 7,
    [types_1.DydxAsset.BNT]: 6,
    [types_1.DydxAsset.MIR]: 6,
    [types_1.DydxAsset.SRM]: 6,
    [types_1.DydxAsset.LON]: 6,
    [types_1.DydxAsset.DODO]: 6,
    [types_1.DydxAsset.ALPHA]: 5,
    [types_1.DydxAsset.WNXM]: 7,
    [types_1.DydxAsset.XCH]: 8,
};
exports.COLLATERAL_ASSET_ADDRESS_BY_NETWORK = {
    [types_1.NetworkId.MAINNET]: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
    [types_1.NetworkId.ROPSTEN]: '0x8707a5bf4c2842d46b31a405ba41b858c0f876c4',
    [types_1.NetworkId.GOERLI]: '0xf7a2fa2c2025ffe64427dd40dc190d47ecc8b36e',
};
exports.COLLATERAL_ASSET_ID_BY_NETWORK_ID = lodash_1.default.mapValues(exports.COLLATERAL_ASSET_ADDRESS_BY_NETWORK, (address) => makeCollateralAssetId(address));
/**
 * Mapping from a synthetic asset to its asset ID.
 */
exports.SYNTHETIC_ASSET_ID_MAP = lodash_1.default.chain(exports.SYNTHETIC_ASSETS)
    .keyBy()
    .mapValues(makeSyntheticAssetId)
    .value();
/**
 * The smallest unit of the asset in the Starkware system, represented in canonical (human) units.
 */
exports.ASSET_QUANTUM_SIZE = lodash_1.default.mapValues(exports.ASSET_RESOLUTION, (resolution) => `1e-${resolution}`);
/**
 * Construct the asset ID (as a 0x-prefixed hex string) for the collateral asset, given the address.
 */
function makeCollateralAssetId(tokenAddress, quantization = 1) {
    const data = Buffer.concat([
        (0, keccak_1.keccak256)(Buffer.from('ERC20Token(address)')).slice(0, 4),
        Buffer.from((0, util_1.normalizeHex32)(tokenAddress), 'hex'),
        Buffer.from((0, util_1.normalizeHex32)(new bn_js_1.default(quantization).toString(16)), 'hex'),
    ]);
    const result = (0, keccak_1.keccak256)(data);
    const resultBN = new bn_js_1.default(result.toString('hex'), 16);
    resultBN.imaskn(250);
    return `0x${(0, util_1.normalizeHex32)(resultBN.toString(16))}`;
}
/**
 * Construct the asset ID (as a 0x-prefixed hex string) for a given synthetic asset.
 */
function makeSyntheticAssetId(asset) {
    const assetIdString = `${asset}-${exports.ASSET_RESOLUTION[asset]}`;
    const assetIdHex = Buffer.from(assetIdString).toString('hex').padEnd(30, '0');
    return `0x${assetIdHex}`;
}
