// @dydxprotocol/starkex-eth/build/src/clients/zeroEx.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateSlippage = exports.getZeroExSwapQuote = void 0;
const big_js_1 = __importDefault(require("big.js"));
const axios_1 = require("../lib/axios/axios");
const request_helpers_1 = require("../lib/axios/request-helpers");
const types_1 = require("../types");
const zeroExUrlMap = {
    [types_1.Networks.MAINNET]: 'https://api.0x.org/swap/v1/quote',
    [types_1.Networks.ROPSTEN]: 'https://ropsten.api.0x.org/swap/v1/quote',
    [types_1.Networks.GOERLI]: 'https://goerli.api.0x.org/swap/v1/quote',
};
async function getZeroExSwapQuote({ sellAmount, sellToken, buyTokenAddress, slippagePercentage, networkId, }) {
    return (0, axios_1.axiosRequest)({
        method: 'GET',
        url: (0, request_helpers_1.generateQueryPath)(zeroExUrlMap[networkId], {
            affiliateAddress: '0xB03fc94a3c49B3126A4E6523D10b65d82C44729C',
            sellAmount,
            sellToken,
            buyToken: buyTokenAddress,
            slippagePercentage,
        }),
    });
}
exports.getZeroExSwapQuote = getZeroExSwapQuote;
function validateSlippage(slippage) {
    if (slippage) {
        const slippageBig = (0, big_js_1.default)(slippage);
        if (slippageBig.lt(0) || slippageBig.gt(1)) {
            throw Error(`Slippage: ${slippage} is not a valid fraction from 0 to 1`);
        }
    }
}
exports.validateSlippage = validateSlippage;
