// @dydxprotocol/v3-client/build/src/eth-signing/signer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Signer = void 0;
const es6_promisify_1 = require("es6-promisify");
const web3_1 = __importDefault(require("web3"));
const types_1 = require("../types");
const helpers_1 = require("./helpers");
class Signer {
    // ============ Constructor ============
    constructor(web3) {
        this.web3 = web3;
    }
    // ============ Functions ============
    /**
     * Returns a signable EIP712 Hash of a struct
     */
    getEIP712Hash(structHash) {
        const hash = web3_1.default.utils.soliditySha3({ t: 'bytes2', v: '0x1901' }, { t: 'bytes32', v: this.getDomainHash() }, { t: 'bytes32', v: structHash });
        // Non-null assertion operator is safe, hash is null only on empty input.
        return hash;
    }
    async ethSignTypedDataInternal(signer, data, signingMethod) {
        let rpcMethod;
        let rpcData;
        let provider = this.web3.currentProvider;
        if (provider === null) {
            throw new Error('Cannot sign since Web3 currentProvider is null');
        }
        if (typeof provider === 'string') {
            throw new Error('Cannot sign since Web3 currentProvider is a string');
        }
        provider = provider;
        let sendAsync;
        switch (signingMethod) {
            case types_1.SigningMethod.TypedData:
                sendAsync = (0, es6_promisify_1.promisify)(provider.send).bind(provider);
                rpcMethod = 'eth_signTypedData';
                rpcData = data;
                break;
            case types_1.SigningMethod.MetaMask:
                sendAsync = (0, es6_promisify_1.promisify)(provider.sendAsync).bind(provider);
                rpcMethod = 'eth_signTypedData_v3';
                rpcData = JSON.stringify(data);
                break;
            case types_1.SigningMethod.MetaMaskLatest:
                sendAsync = (0, es6_promisify_1.promisify)(provider.sendAsync).bind(provider);
                rpcMethod = 'eth_signTypedData_v4';
                rpcData = JSON.stringify(data);
                break;
            case types_1.SigningMethod.CoinbaseWallet:
                sendAsync = (0, es6_promisify_1.promisify)(provider.sendAsync).bind(provider);
                rpcMethod = 'eth_signTypedData_v4';
                rpcData = data;
                break;
            default:
                throw new Error(`Invalid signing method ${signingMethod}`);
        }
        const response = await sendAsync({
            method: rpcMethod,
            params: [signer, rpcData],
            jsonrpc: '2.0',
            id: Date.now(),
        });
        if (response.error) {
            throw new Error(response.error.message);
        }
        return `0x${(0, helpers_1.stripHexPrefix)(response.result)}0${types_1.SignatureTypes.NO_PREPEND}`;
    }
    /**
     * Sign a message with `personal_sign`.
     */
    async ethSignPersonalInternal(signer, message) {
        let provider = this.web3.currentProvider;
        if (provider === null) {
            throw new Error('Cannot sign since Web3 currentProvider is null');
        }
        if (typeof provider === 'string') {
            throw new Error('Cannot sign since Web3 currentProvider is a string');
        }
        provider = provider;
        const sendAsync = ((0, es6_promisify_1.promisify)(provider.sendAsync || provider.send).bind(provider));
        const rpcMethod = 'personal_sign';
        const response = await sendAsync({
            method: rpcMethod,
            params: [signer, message],
            jsonrpc: '2.0',
            id: Date.now(),
        });
        if (response.error) {
            throw new Error(response.error.message);
        }
        // Note: Using createTypedSignature() fixes the signature `v` value.
        return (0, helpers_1.createTypedSignature)(response.result, types_1.SignatureTypes.PERSONAL);
    }
}
exports.Signer = Signer;
