// @dydxprotocol/starkex-eth/build/__tests__/StarkwareLib.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_1 = __importDefault(require("web3"));
const usdc_abi_json_1 = __importDefault(require("../src/contracts/usdc-abi.json"));
const index_1 = require("../src/index");
const types_1 = require("../src/types");
const DummyProvider_1 = require("./helpers/DummyProvider");
describe('StarwareLib#initiate', () => {
    const dummyProvider = new DummyProvider_1.DummyProvider();
    it('Successfully initiates StarkwareLib on Goerli', async () => {
        const starkwareLib = new index_1.StarkwareLib(dummyProvider, types_1.Networks.GOERLI);
        expect(starkwareLib.collateralToken).toBeTruthy();
        expect(starkwareLib.contracts).toBeTruthy();
        expect(starkwareLib.exchange).toBeTruthy();
        expect(starkwareLib.factRegistry).toBeTruthy();
        expect(starkwareLib.logs).toBeTruthy();
        expect(starkwareLib.mintableToken).toBeTruthy();
        try {
            const realProvider = new web3_1.default.providers.HttpProvider(process.env.ETHEREUM_NODE_RPC_URL, { timeout: 10000 });
            const real = new index_1.StarkwareLib(realProvider, types_1.Networks.GOERLI);
            const account = real.web3.eth.accounts.privateKeyToAccount(process.env.PRIVATE_KEY);
            real.web3.eth.accounts.wallet.add(account);
            real.web3.eth.defaultAccount = account.address;
            await real.exchange.setERC20Allowance({
                tokenAddress: usdc_abi_json_1.default.networks[5].address,
                address: '0x029bB89d64695D6A461eEbC1Aab4a4C8657a3f22',
                amount: 100,
            }, { from: account.address });
            const result = await real.exchange.getERC20Allowance({
                ownerAddress: account.address,
                tokenAddress: usdc_abi_json_1.default.networks[5].address,
                spenderAddress: '0x029bB89d64695D6A461eEbC1Aab4a4C8657a3f22',
                decimals: 8,
            });
            expect(result).toEqual('0.000001');
        }
        catch (error) {
            // eslint-disable-next-line no-console
            console.error(error);
        }
    });
});
