// @dydxprotocol/v3-client/build/__tests__/api-keys.test.js
"use strict";
/**
 * Unit tests for the API keys module.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const web3_1 = __importDefault(require("web3"));
const src_1 = require("../src");
const util_1 = require("./helpers/util");
const apiKeyCredentials = {
    key: 'd53c3a7d-3add-68db-a9c3-9ad582313c8e',
    secret: '85BR_H-GC7HS3aydOxLw3zjRuDI6RYVgFmsYaKJh',
    passphrase: '1qYatmED3wy9RnDZsGnR',
};
describe('API Keys Module & Private Module', () => {
    it('signs a private request', async () => {
        (0, util_1.asMock)(axios_1.default).mockResolvedValue({});
        const web3 = new web3_1.default();
        const client = new src_1.DydxClient('https://example.com', { web3, apiKeyCredentials });
        await client.private.getApiKeys();
        expect(axios_1.default).toHaveBeenCalledTimes(1);
        expect(axios_1.default).toHaveBeenCalledWith({
            url: expect.stringContaining('/v3/api-keys'),
            method: 'GET',
            headers: {
                'DYDX-API-KEY': expect.stringMatching(/[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}/),
                'DYDX-TIMESTAMP': expect.any(String),
                'DYDX-PASSPHRASE': expect.stringMatching(/^[A-Za-z0-9_-]{20}$/),
                'DYDX-SIGNATURE': expect.any(String),
            },
            data: undefined,
        });
    });
    it('signs an ApiKey request', async () => {
        (0, util_1.asMock)(axios_1.default).mockResolvedValue({});
        const web3 = new web3_1.default();
        const account = web3.eth.accounts.wallet.create(1)[0];
        const client = new src_1.DydxClient('https://example.com', { web3 });
        await client.ethPrivate.deleteApiKey(apiKeyCredentials.key, account.address);
        expect(axios_1.default).toHaveBeenCalledTimes(1);
        expect(axios_1.default).toHaveBeenCalledWith({
            url: expect.stringContaining('/v3/api-keys'),
            method: 'DELETE',
            headers: {
                'DYDX-SIGNATURE': expect.stringMatching(/0x[0-9a-f]{130}/),
                'DYDX-TIMESTAMP': expect.any(String),
                'DYDX-ETHEREUM-ADDRESS': expect.stringMatching(/0x[0-9a-fA-F]{40}/),
            },
        });
    });
});
